### DISCUSSION TITLE
Multiple language Static pages explanation/help

### DISCUSSION THREAD

**Discussion by levipadre** [vercel/next.js#54467] [Help]
I need some explanation from an expert, who understands Next.js well.

### Some info about the project:
I have a website with thousand of pages on different domain names, as it's international.
The domain names formats are `domain.com`, `domain.co.uk`, `domain.fr`, `domain.de` etc.

Every pages are coming from a CMS through a RestAPI. Because there are multiples different page types and any pages can be anywhere, I thought way to achieve is with Dynamic Router/Optional Catch-all Segments (https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes#optional-catch-all-segments):

My structure is then:`app/[[...slug]]/page`

### What I want to achive:
I want to generate static pages, caching them and revalidate them on-demand, but the pages are server side generated. Accoring to this page about Caching (https://nextjs.org/docs/app/building-your-application/caching#revalidating-1) it should be possible. Even 'CMS' word has 
 mentioned.

### The problem:
I can't make the pages work if I don't have the host or locale, bacause when I call a page with slug, e.g. `/about-us`, it can belong to `domain.com` or `domain.co.uk` of course.

So:
- if I use middleware to recognise the domain (e.g. `app/[lang]/[[...slug]]/page`), it becomes SSR
- if I use `headers` to get the host, it becomes SSR

### The question:
How can I make them static pages but still know the locale/domain/host?

#### Code:
My `app/[[...slug]]/page` code ATM using `headers`. It works, but unfortunately it's SSR:

```
import { getPageByUri } from '@/lib/APIs/pages';
import { headers } from 'next/headers';
import { extractTLD } from '@/lib/extractTLD';

export default async function Page({
    params: {
        slug
    }
}: {
    params: {
        slug: string[]
    }
}) {
    const locale = extractTLD(headers().get('x-forwarded-host') || headers().get('host'));
    const path = (typeof slug === 'undefined' || slug[0] === 'index') ? '' : slug.join('/');;
    const page = await getPageByUri(path, locale);

    return (
        <>
            <h1>{page.title}</h1>
        </>
    )
}
```

### Additional information

```js
Nextjs version: 13.4.19
Node version: 18.17
```

**Comment by levipadre** [comment]
Someone please!

**Reply by flextremedev** [reply]
This doesn't help right?
https://nextjs.org/docs/app/building-your-application/routing/internationalization#static-generation

**Comment by joaopedrodcf** [comment]
Hey @levipadre I have a possible solution for you it will come with a caveat

You could create each route with the /[lang]/ as you can read below:
https://next-intl-docs.vercel.app/docs/routing/middleware#usage-without-middleware-static-export

I know it will look super weird as you already have different domains per language and would endue being domain.de/de

But this is the only way I can see of having different assets per language without the need of a server. 

This still needs some change in for example Cloudflare for you to point the domain.de to /de with redirects.

**Reply by joaopedrodcf** [reply]
Also you can try to use https://react.i18next.com but this will make it lose the purpose of having a SSG application where all the routes were already created previously :(

**Reply by levipadre** [reply]
Hi @joaopedrodcf, thank you for your reply. I’m not sure I fully understand your suggested solution, though.

- will I be able to revalidate on-demand if I use static export as you suggested?
- I don't only have different domains per language, but sometimes multiple domains per language. For example domain.com, domain.uk and domain.eu are both English, but with different content from the CMS

You don't have an example code by any chance, do you?

**Comment by GabenGar** [comment]
Does your app work across different domains or what?

**Reply by levipadre** [reply]
Yes, it does. E.g. domain.com, domain.co.uk, domain.fr, domain.de etc.

**Reply by GabenGar** [reply]
How do you decide which value CMS returns belongs to which domain?

**Reply by levipadre** [reply]
Because I can't use locale, like /en/foo/bar or /es/uno/dos to get the page data, I have to use /united-kingdom or /united-states or /espana. So obviously, it's a change in the CMS to provide the right endpoint. But after that I just get the host info from headers and change the slug based on that, like:
```
export async function getPageByUri(uri: string, locale: string) {
    switch (locale.substring(0, 2)) {
    case 'uk':
        locale = 'united-kingdom';
        break;
    case 'com':
        locale = 'united-states';
        break;
    case 'org':
        locale = 'global';
        break;
    case 'fr':
        locale = 'france';
        break;
    case 'it':
        locale = 'italia';
        break;
    case 'nl':
        locale = 'nederland';
        break;
    case 'pl':
        locale = 'polska';
        break;
    case 'se':
        locale = 'sverige';
        break;
    case 'cz':
        locale = 'ceska-republika';
        break;
    case 'de':
        locale = 'deutschland';
        break;
    case 'es':
        locale = 'espana';
        break;
    case 'cn':
        locale = 'china';
        break;
    }

    const data = await fetchAPI(`/umbraco/delivery/api/v1/content/item/${locale}/${uri}`, {
        method: 'GET',
    });
    return data;
}
```

My problem is that I have to use headers, so they can't be static pages. It's a shame that we have to stick with the exact url to be able to get static pages.

**Reply by GabenGar** [reply]
These are not valid BCP 47 tags btw. And you still didn't explain how do you plan to feed the static pages to all those quadrillion domains, because nextjs is not going to do that for you.

**Reply by levipadre** [reply]
No, they are not. All custom, because we have 3-4 English sites that we can't differentiate otherwise. So it's not a typical multilanguage project, but a multisite project.

I thought I can get the static pages by `generateStaticParams`like:
```
export async function generateStaticParams() {
    const pages = await fetch('/api/pages').then((res) => res.json())

    return pages.map((page: any) => ({
        slug: [page.url]
    }))
}
``` 
but I either need to use header to get the host or middleware. Both of them will turn the site into SSR.

**Reply by martink635** [reply]
@levipadre Did you manage to find a solution for this?
